[ONNX] Introduce 'Functionalization' for fx exporter

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98246
* __->__ #98245

<img src="https://user-images.githubusercontent.com/9376104/229648898-7e85efc8-143f-42f9-93e0-298a8f86c0a1.png" width="80%" height="80%">

BowenBao committed:
[ONNX] Introduce 'Functionalization' for fx exporter

[ghstack-poisoned]

BowenBao commented on test/onnx/test_fx_dynamic_with_onnxruntime.py:
```diff
@@ -196,6 +196,18 @@ def forward(self, x, y):
             DynamicAdd(), (x, y), additional_test_inputs=[(input_x, input_y)]
         )
 
+    def test_mutation(self):
```
To be deduplicated once dynamic tests are parameterized.

BowenBao committed:
Update on "[ONNX] Introduce 'Functionalization' for fx exporter"




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Introduce 'Functionalization' for fx exporter"




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Introduce 'Functionalization' for fx exporter"


<img src="https://user-images.githubusercontent.com/9376104/229648898-7e85efc8-143f-42f9-93e0-298a8f86c0a1.png" width="80%" height="80%">


[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/fx_exporter.py:
```diff
@@ -2,6 +2,7 @@
 from typing import Any, Callable, Mapping, Sequence
 
 import torch._ops
+import torch.fx
```
for intellisense/pylance not complaining `"fx" is not a known member of module "torch"` for `torch.fx.GraphModule`.

justinchuby commented on torch/onnx/_internal/fx/passes/utils.py:
```diff
@@ -0,0 +1,53 @@
+from __future__ import annotations
```
Should we limit the scope of this utils module by adding explicit guidance here?

BowenBao replied:
wdym? could you elaborate

justinchuby replied:
Right it was about the module. Sorry for the confusion

BowenBao replied:
In addition to docstring, what would be a better file name, `utils.py`, `common.py`, `common_utils.py`, etc?

justinchuby replied:
I think we can import passes and use `passes.utils`. `passes.common` sounds great too

BowenBao replied:
I'll go with `passes._utils` then, they are not supposed to be used outside lol.

justinchuby reviewed (commented):
Let's go!!

wschin commented on torch/onnx/_internal/fx/passes/functionalization.py:
```diff
@@ -0,0 +1,79 @@
+    * ``torch.func.functionalize`` also removes mutations (and views) that were performed
+    on function inputs. However to preserve semantics, functionalize will "fix up" the
+    mutations after the transform has finished running, by detecting if any tensor inputs
+    "should have" been mutated, and copying the new data back to the inputs if necessary.
```
How does `copying the new data back to the inputs` mean?

BowenBao replied:
Consider
```python
def fn(a, b):
    a.add_(b)
    return a
```
When you call this function `fn(x, y)`, `x` is also mutated. So just functionalizing is not enough for preserving the original semantics. A "special" input mutation step, if needed, needs to be inserted at the end. That's what the comment means. 
```python
# after functionalization
# this is not semantically the same. `a`'s value is not updated.
def fn(a, b):
    new_a = a + b
    return new_a

# functionalization that preserves semantics
def fn(a, b):
    new_a = a + b

    # copying the new data back to the inputs
    a.copy_(new_a)

    return new_a
```

BowenBao replied:
added above into docstring

wschin commented on torch/onnx/_internal/fx/passes/utils.py:
```diff
@@ -0,0 +1,53 @@
+            f"module: {len(placeholders)}, reference_module: {len(reference_placeholders)}"
+        )
+
+    for placeholder, reference_placeholder in zip(placeholders, reference_placeholders):
```
Let's handle potential name conflicts here.
```
def foo(x1, y1):
  x=x1+y1
  b=a*x1 
```
and
```
def foo_reference(x, y):
  a=x+y
  b=a*x 
```

Either
1. Assert if any names from `reference_module` are being used.
2. Change the name `x` used in `foo` to `x_another` and then you can replace `x1` with `x`.

BowenBao replied:
Good catch, this function is unsafe. I was just moving this function in this PR. For the sake of functionalization scoping, I'm happy to expand on this in another PR.

BowenBao replied:
added above into docstring

wschin replied:
Another PR is fine. Just want to note that at large models such as GPT, name conflict can easily occur. :)

BowenBao committed:
Update on "[ONNX] Introduce 'Functionalization' for fx exporter"


<img src="https://user-images.githubusercontent.com/9376104/229648898-7e85efc8-143f-42f9-93e0-298a8f86c0a1.png" width="80%" height="80%">


[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/passes/functionalization.py:
```diff
@@ -0,0 +1,104 @@
+
+    @_beartype.beartype
+    def _run(self, *args) -> torch.fx.GraphModule:
+        for node in reversed(self.module.graph.nodes):
```
Why do we need reversed iterator?

BowenBao replied:
Here one of the conditions to meet when deleting a node is it is unused by others. The graph is topologically sorted, meaning user node always comes after used node. So deleting from backwards ensures to delete the nodes whose users are all deleted.

titaiwangms commented on torch/onnx/_internal/fx/passes/functionalization.py:
```diff
@@ -0,0 +1,104 @@
+
+    For ONNX inference, it is recommended to run ``RemoveInputMutation`` after this pass.
+    ``RemoveInputMutation`` removes the "fix up" nodes that were added by ``Functionalize``,
+    which are not needed for ONNX inference.
```
Why not remove fix up in this pass?

BowenBao replied:
Because input mutation might be useful in training.

titaiwangms replied:
So we are likely to bypass the RemoveInputMutation pass if model is in training mode?

BowenBao replied:
Yes.

To add more details, I think how to make use of this information depends on how `dynamo_export` integrates with training. But generally speaking, it is always better to preserve as much information as possible.

titaiwangms commented on torch/onnx/_internal/fx/fx_exporter.py:
```diff
@@ -24,21 +25,27 @@ def export_fx_to_onnx(
             enable_dynamic_axes=self.options.dynamic_shapes,
         ).run(*fx_module_args)
 
+        # Resolve mutations.
```
Can we have an explanation why we need to resolve mutations somewhere in the codebase? It's for future maintenance.

BowenBao replied:
There is a test case added. It fails otherwise.

BowenBao replied:
I misunderstood your comment... Sure.

BowenBao replied:
```
        # ONNX does not support views and mutations.
        # Functionalize to get a semantically equivalent graph without mutations.
```
wdyt?

titaiwangms replied:
Thanks!

titaiwangms reviewed (approved):
LGTM!

BowenBao committed:
remove unused code; add comment on "[ONNX] Introduce 'Functionalization' for fx exporter"


<img src="https://user-images.githubusercontent.com/9376104/229648898-7e85efc8-143f-42f9-93e0-298a8f86c0a1.png" width="80%" height="80%">


[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge
